controller/userController.js

```diff
@@ -114,157 +103,169 @@ const create = (req, res) => {
         return res.status(HttpStatus.BAD_REQUEST).json({
-            status: StatusEnum['ERROR'],
             message: 'Invalid ID'
         });
     }
 
+    const userId = req.params.id;
+    const loggedUser = req.decoded;
+
     User.findById(userId, { passwordHash: 0, salt: 0 }).then(user => {
         // Client can only retrieve users within his company
         if (user && user.companyID === loggedUser.companyID) {
             return res.status(HttpStatus.OK).json(user);
         } else {
             return res.status(HttpStatus.NOT_FOUND).json({
-                status: StatusEnum['ERROR'],
                 message: 'User not found.'
             });
         }
     }).catch(err => {
         return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
-            status: StatusEnum['ERROR'],
             message: ValidationUtil.buildErrorMessage(err, 'retrieveById', 'user')
         });
     });
 };
 
-const update = (req, res) => {
+const changePersonalData = (req, res) => {
     const loggedUser = req.decoded;
 
     if (!ValidationUtil.isValidObjectId(loggedUser._id)) {
-        return res.status(HttpStatus.BAD_REQUEST).json({
-            status: StatusEnum['ERROR'],
-            message: 'Invalid ID'
-        });
+        return res.status(HttpStatus.BAD_REQUEST).json({ message: 'Invalid ID' });
     }
-    user = {};
-    // TODO: 
-    // Make 4 functions from this one:
-    //  1. updateFields() (everything except password & active flag)
-    //  2. updatePassword()
-    //  3. blockUser (basically remove)
-    //  3. activateUser
+
+    let user = {};
     // Update fields
     if (req.body.name)
         user.name = req.body.name;
     if (req.body.surname)
         user.surname = req.body.surname;
-    if (req.body.email)
-        user.email = req.body.email;
     if (req.body.dateOfBirth)
         user.dateOfBirth = req.body.dateOfBirth;
-    if (req.body.active)
-        user.active = req.body.active;
+
+    User.findOneAndUpdate({ _id: loggedUser._id }, user, { useFindAndModify: false, new: true, runValidators: true }).then(newUser => {
+        if (newUser) {
+            // Generate new JWT
+            const token = jwt.sign(JSON.stringify(newUser), AppConfig.SECRET);
+            newUser.passwordHash = null;
+            newUser.salt = null;
+
+            return res.status(HttpStatus.OK).json({ token, user: newUser });
+        } else {
+            return res.status(HttpStatus.NOT_FOUND).json({
+                message: 'User not found.'
+            });
+        }
+    }).catch(err => {
+        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
+            message: 'Internal server error'
+        });
+    });
+};
+
+const changePassword = (req, res) => {
+    const loggedUser = req.decoded;
+
+    if (!ValidationUtil.isValidObjectId(loggedUser._id)) {
+        return res.status(HttpStatus.BAD_REQUEST).json({ message: 'Invalid ID' });
```

Comment:
We should discuss these return messages in cases like this one. What's the real use of a message like 'Invalid ID'? Are we going to show it to the app user? Are we going to string compare some hard coded values on user app to this message? @SavoX2 @djuricic1 what are your thoughts on this? Looking at my code I too used not so descriptive messages, so I think we should reconsider the use of these messages and discuss what we want to do with them.

Reply:
Also since `loggedUser` is taken from the header token I don't see a reason to check if the _id is valid since if it weren't the decoding would fail. I also believe that the method itself takes a `string` as a parameter, where this is an actual `Mongoose.Types.ObjectId`.